src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+        //       something that is not allowed in CellML and will therefore be
+        //       reported when we validate the model.
+
+        // Have our equation track the (ODE) variable (by ODE variable, we mean
+        // a variable that is used in a "diff" element).
+
+        if (node->parent()->firstChild()->isMathmlElement("diff")) {
+            equation->addOdeVariable(analyserVariable(variable));
+        } else if (!(node->parent()->isMathmlElement("bvar")
+                     && node->parent()->parent()->firstChild()->isMathmlElement("diff"))) {
+            equation->addVariable(analyserVariable(variable));
+        }
+
+        // Add the variable to our AST.
+
+        ast->mPimpl->populate(AnalyserEquationAst::Type::CI, variable, astParent);
+    } else if (node->isMathmlElement("cn")) {
+        if (mathmlChildCount(node) == 1) {
+            // We are dealing with an e-notation based CN value.
+
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString() + "e" + node->firstChild()->next()->next()->convertToString(), astParent);
+        } else {
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString(), astParent);
+        }
+
+        // Qualifier elements.
+
+    } else if (node->isMathmlElement("degree")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::DEGREE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+    } else if (node->isMathmlElement("logbase")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::LOGBASE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+    } else if (node->isMathmlElement("bvar")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::BVAR, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+
+        auto rightNode = mathmlChildNode(node, 1);
+
+        if (rightNode != nullptr) {
+            processNode(rightNode, ast->mPimpl->mRightChild, ast, component, equation);
+        }
+
+        // Constants.
+
+    } else if (node->isMathmlElement("true")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::TRUE, astParent);
+    } else if (node->isMathmlElement("false")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::FALSE, astParent);
+    } else if (node->isMathmlElement("exponentiale")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::E, astParent);
+    } else if (node->isMathmlElement("pi")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::PI, astParent);
+    } else if (node->isMathmlElement("infinity")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::INF, astParent);
+    } else if (node->isMathmlElement("notanumber")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::NAN, astParent);
+    }
```

Comment:
The spec lists a `<sep>` element that's not tested here? https://cellml-specification.readthedocs.io/en/latest/reference/formal_and_informative/specB12.html#id1

Reply:
We indirectly deal with it here (the e-notation case):
```
    } else if (node->isMathmlElement("cn")) {
        if (mathmlChildCount(node) == 1) {
            // We are dealing with an e-notation based CN value.

            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString() + "e" + node->firstChild()->next()->next()->convertToString(), astParent);
        } else {
            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString(), astParent);
        }
```